functions.php

```diff
@@ -0,0 +1,270 @@
+<?php
+
+function toJson(array $data): ?string
+{
+    try {
+        $jsonEncoded = json_encode($data, JSON_ERROR_SYNTAX, 512);
+    } catch (JsonException $e) {
+        echo $e->getMessage();
+        return null;
+    }
+
+    return $jsonEncoded;
+}
+
+function fromJson(string $data): ?array
+{
+    try {
+        $jsonDecoded = json_decode($data, true, 512, JSON_ERROR_SYNTAX);
+    } catch (JsonException $e) {
+        echo $e->getMessage();
+        return null;
+    }
+
+    return $jsonDecoded;
+}
+
+function sendResponse(string $message, int $statusCode): void
+{
+    http_response_code($statusCode);
+
+    echo $message;
+
+    exit;
+}
+
+function processURL(string &$endpoint, array &$params, string &$method): void
+{
+    $url = parse_url($_SERVER['REQUEST_URI']);
+
+    $endpoint = str_replace('/', '', $url['path']);
+
+    $params = [];
+
+    if (isset($url['query'])) {
+        parse_str($url['query'], $params);
+    }
+
+    $method = $_SERVER['REQUEST_METHOD'];
+}
+
+function readProducts(string $path): ?array
+{
+    $productsRead = @file_get_contents($path);
+
+    if ($productsRead === false) {
+        return null;
+    }
+
+    return fromJson($productsRead);
+}
+
+function extractProduct(array $products, string $category, ?int $id = null): ?array
+{
+    if (!isset($products[$category])) {
+        return null;
+    }
+
+    if ($id !== null) {
+        foreach ($products[$category] as $product) {
+            if ($product['id'] == $id) {
+                return $product;
+            }
+        }
+
+        return null;
+    }
+
+    return $products[$category];
+}
+
+function addToList(array &$products, array $requestData, ?array $oldProduct = null, bool $newId = false): void
+{
+    $fields = [ID, NAME, PRICE, QUANTITY];
+
+    foreach ($fields as $property) {
+        $updatedProduct[$property] = (
+            isset($requestData[$property]) ? $requestData[$property] :
+            (isset($oldProduct[$property]) ? $oldProduct[$property] : null)
+        );
```

Comment:
completely unreadable

Reply:
@AndreiOrmanji - Achieved a more readable syntax by using separate ternary operators to check property in each parameter.